File: apple/Sources/MeshDropKit/TXTRecord.swift
import Foundation
import Network

/// mDNS TXT 记录的编解码。字段见 [discovery.md](../../../protocol/discovery.md)。
public enum TXTRecord {
    public static let serviceType = "_meshdrop._tcp"

    public static func encode(
        identity: Identity,
        displayName: String,
        os: DeviceOS,
        model: String?,
        port: UInt16,
        protocolVersion: UInt8 = 1
    ) -> NWTXTRecord {
        var record = NWTXTRecord()
        record["v"] = String(protocolVersion)
        record["id"] = identity.id
        record["name"] = base64URLEncode(displayName.data(using: .utf8) ?? Data())
        record["os"] = os.rawValue
        if let model { record["model"] = model }
        record["fp"] = identity.fingerprint
        record["port"] = String(port)
        return record
    }

    /// 解析 TXT。任何必选字段缺失或格式错误返回 nil（调用方应当忽略此服务，
    /// 不能视为协议错误 — 同网段可能存在新版本广告的旧端解析不了的字段）。
    public static func decode(_ record: NWTXTRecord) -> Device? {
        guard let v = record["v"], let versionInt = UInt8(v),
              let id = record["id"], id.count == 32,
              let nameB64 = record["name"],
              let nameData = base64URLDecode(nameB64),
              let name = String(data: nameData, encoding: .utf8),
              let osStr = record["os"], let os = DeviceOS(rawValue: osStr),
              let fp = record["fp"], fp.count == 32,
              let portStr = record["port"], let port = UInt16(portStr)
        else {
            return nil
        }
        return Device(
            id: id,
            name: name,
            os: os,
            model: record["model"],
            fingerprint: fp,
            port: port,
            protocolVersion: versionInt
        )
    }

    // MARK: - base64url (no padding)

    static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    static func base64URLDecode(_ string: String) -> Data? {
        var s = string
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let pad = (4 - s.count % 4) % 4
        s.append(String(repeating: "=", count: pad))
        return Data(base64Encoded: s)
    }
}

File: windows/ShareWindows/Models/TXTRecord.cs
using System.Collections.Generic;
using System.Text;

namespace MeshDrop.Models;

/// <summary>
/// mDNS TXT 记录的编解码。字段见 protocol/discovery.md。
/// </summary>
public static class TXTRecord
{
    public const string ServiceType = "_meshdrop._tcp";

    public static Dictionary<string, string> Encode(
        Identity identity,
        string displayName,
        DeviceOS os,
        string? model,
        ushort port,
        byte protocolVersion = 1)
    {
        var dict = new Dictionary<string, string>
        {
            ["v"] = protocolVersion.ToString(),
            ["id"] = identity.Id,
            ["name"] = Base64Url.Encode(Encoding.UTF8.GetBytes(displayName)),
            ["os"] = os.ToRaw(),
            ["fp"] = identity.Fingerprint,
            ["port"] = port.ToString(),
        };
        if (!string.IsNullOrEmpty(model)) dict["model"] = model;
        return dict;
    }

    public static Device? Decode(IReadOnlyDictionary<string, string> attrs)
    {
        if (!attrs.TryGetValue("v", out var v) || !byte.TryParse(v, out var version)) return null;
        if (!attrs.TryGetValue("id", out var id) || id.Length != 32) return null;
        if (!attrs.TryGetValue("name", out var nameB64)) return null;
        var nameBytes = Base64Url.Decode(nameB64);
        if (nameBytes is null) return null;
        var name = Encoding.UTF8.GetString(nameBytes);
        if (!attrs.TryGetValue("os", out var osStr)) return null;
        var os = DeviceOSExtensions.Parse(osStr);
        if (os is null) return null;
        if (!attrs.TryGetValue("fp", out var fp) || fp.Length != 32) return null;
        if (!attrs.TryGetValue("port", out var portStr) || !ushort.TryParse(portStr, out var port)) return null;
        attrs.TryGetValue("model", out var model);
        return new Device(id, name, os.Value, model, fp, port, version);
    }
}

internal static class Base64Url
{
    public static string Encode(byte[] data)
    {
        var s = System.Convert.ToBase64String(data);
        return s.Replace('+', '-').Replace('/', '_').TrimEnd('=');
    }

    public static byte[]? Decode(string s)
    {
        var normalized = s.Replace('-', '+').Replace('_', '/');
        var pad = (4 - normalized.Length % 4) % 4;
        normalized = normalized.PadRight(normalized.Length + pad, '=');
        try { return System.Convert.FromBase64String(normalized); }
        catch { return null; }
    }
}
